keps/sig-api-machinery/2887-openapi-enum-types/README.md

```diff
@@ -0,0 +1,836 @@
+marker to type aliases that represent enums. The OpenAPI generator will have
+the capability to recognize the marker and auto-detect possible values for an enum.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, types that are actually enums are represented as plain strings. For example,
+here is the definition of `Protocol`, which must be one of `TCP`, `UDP`, and `SCTP`.
+```go
+package core
+// ...
+
+// Protocol defines network protocols supported for things like container ports.
+type Protocol string
+
+const (
+	// ProtocolTCP is the TCP protocol.
+	ProtocolTCP Protocol = "TCP"
+	// ProtocolUDP is the UDP protocol.
+	ProtocolUDP Protocol = "UDP"
+	// ProtocolSCTP is the SCTP protocol.
+	ProtocolSCTP Protocol = "SCTP"
+)
+
+// ContainerPort represents a network port in a single container
+type ContainerPort struct {
+  // ...
+
+  // Required: Supports "TCP", "UDP" and "SCTP"
+  // +optional
+  Protocol Protocol
+  
+  // ...
+}
+```
+
+In order to convey that the `Protocol` type is an enum, every field of the type 
+have to comment its values like `ContainerPort` does. The same goes for validation - 
+protocol being one of "TCP", "UDP", and "SCTP" is duplicated among the code base.
+
+Tools and UIs can also benefit from the additional piece of information. For example,
+Protocol can be shown as a dropdown of possible values instead of a textbox.
+The user no longer needs to see an error of invalid value after submits the request.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add a marker that indicates a field as an enum.
+- Make the Kubernetes OpenAPI generator recognize the marker.
+- Deduce values of an enum type from its Go type definition.
+- Detect and annotate enum types in all built-in types. 
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+This KEP does not require removal of existing validation of enum types. While OpenAPI-based
+of enum types can simplify this type of validation, retiring custom-made validations is out
+of scope.
+
+This proposal does not intend to change the enum syntax for third-party tools like `kubebuilder`.
+The implementation should only suggest changes to internal APIs. 
+
+## Proposal
+
+When generating OpenAPI schema, the de facto enum types should be reflected as enums in the resulting schema.
+
+We define enum types with following properties:
+- All enum types are closed, i.e., the type includes a finite number of possible values.
```

Comment:
Closed enums are a disaster for forward/backwards compatibility, but this KEP doesn't seem to have a discussion of that; I think it should, if we're going to automatically make all of these enums closed. Currently, generated clients won't represent these with closed sets. This would be a big change.

Reply:
They are closed from a server-side perspective (server won't allow any unknown value). Clients, for backwards compatibility reasons, should consider it open (new values are possible).

Reply:
So, would you suggest always prune the enum fields for clients?

Reply:
I think this is up to the maintainer of a client. We just give options.